add subtypes for chain cells

yusefnapora opened the pull request:
putting this up early to get some feedback.

I split the transactor and datastore types into separate objects in the new `io.mediachain.types` package.   This should make it easier to pull into a separate project later.  `DatastoreTypes` has everything that inherits from `DataObject`, and `TransactorTypes` has the transactor-internal types like `ChainReference` and the journal interfaces.  The `DatastoreTypes` are all `CborSerializable`, but the `TransactorTypes` use pojo serialization.

Also changes the `EntityChainCell` and `ArtefactChainCell` to regular classes and adds some case classes that extend them (only a few so far).

the existing tests pass, but I haven't added specs for unpacking the chain cell subtypes yet.

yusefnapora committed:
split datastore & transactor types into separate objects

yusefnapora committed:
make EntityChainCell & ArtefactChainCell regular classes (not case classes)

yusefnapora committed:
add some chain cell subtypes

vyzo commented on transactor/src/main/scala/io/mediachain/types/Datastore.scala:
```diff
@@ -0,0 +1,8 @@
+  import DatastoreTypes.{DataObject, Reference}
+
+  def put(obj: DataObject): Reference
+}
```
no newline at end of file

vyzo commented on transactor/src/main/scala/io/mediachain/types/DatastoreTypes.scala:
```diff
@@ -0,0 +1,202 @@
+package io.mediachain.types
+
+
+object DatastoreTypes {
```
can't we call this just Datastore and put the Datastore interface together as well?
DatastoreTypes is a very ugly name, and it is also rather sad to have a forlorn Datastore trait with one method in its own file.

yusefnapora replied:
sounds good to me :)

vyzo replied:
it is a member of a package named io.mediachain.types after all, so types.Datastore and types.Transactor is natural.

vyzo commented on transactor/src/main/scala/io/mediachain/types/TransactorTypes.scala:
```diff
@@ -0,0 +1,54 @@
+import scala.concurrent.Future
+
+
+object TransactorTypes {
```
similarly, let's call this just Transactor.

yusefnapora committed:
use specific MediachainType for base chain cells if present, fallback to generic type

yusefnapora committed:
add cell subtypes to sets in MediachainTypes

yusefnapora committed:
rename DatastoreType -> Datastore, TransactorType -> Transactor

vyzo commented:
I like the idea of putting the chain cell types in the meta as a field, it's a flexible approach.

yusefnapora committed:
fix incorrect mediachain type

yusefnapora committed:
add remaining artefact chain cells

yusefnapora committed:
deserializers for chain cell subtypes

yusefnapora committed:
add methods to decode DataObjects and JournalEntries from byte arrays

yusefnapora committed:
comment for DeserializerMap

yusefnapora committed:
cbor specs for EntityChainCell subtypes, override toCbor on cell subtypes

yusefnapora committed:
round-trip cbor specs for aretefact cell subtypes

yusefnapora committed:
test that transactorDeserializers decodes to base cell type

yusefnapora committed:
fix case on filename (DataStore -> Datastore)

bigs commented on transactor/src/main/scala/io/mediachain/types/CborSerialization.scala:
````diff
@@ -1,51 +1,123 @@
+    * val fooCell = fromCborBytes(fooCellBytes)
+    * ```
+    */
   type DeserializerMap = Map[MediachainType, CborDeserializer[CborSerializable]]
````
what if we took the witness approach to this... this way the passing around of `deserializers` wouldn't be so shifty

``` scala
  trait Deserializer[T] {
    def deserialize(bytes: Array[Byte]): Option[T]
  }

  // dumb string deserialization
  implicit val deserializeInt = new Deserializer[Int] {
    override def deserialize(bytes: Array[Byte]): Option[Int] = {
      val str = new String(bytes)
      try {
        str.toInt
      } catch {
        case _ => None
      }
    }
  }

  def foo(myInt: Array[Byte])(implicit d: Deserializer[Int]): Int = {
    d.deserialize(myInt).map(_ + 1).getOrElse(0)
  }

  def generalDeserializer[T](implicit d: Deserializer[T]): Option[T] = {
     val bytes = fetchDb()
     d.deserialize(bytes)
  }
```

bigs replied:
ah realizing this is for reifying types

bigs replied:
ok so this impl is actually fine but i'm not sure that u have to pass around the implicit `derserializers` everywhere. it's really confusing to see it not explicitly passed in to all the fns that accept it as a parameter

parkan replied:
pretty standard, same as the json4s type conversions etc

vyzo commented:
For the new chain cell some classes, we will need to add a cons method that constructs the exact type for use inside the state machine.
I will add it and send a PR on top of this.

vyzo committed:
support chain cell subtypes in state machine

- adds a cons method to ChainCells for consing on top of a chain
- modifies the state machine to use cons for constructing the actual chain cells

vyzo commented:
PR #55 for supporting chain cell subtypes in the state machine.

vyzo committed:
Copycat: remove errant import

yusefnapora committed:
Merge pull request #55 from mediachain/vyzo-chain-types-cons

yn-chain-types: support chain cell subtypes in state machine

vyzo commented on transactor/src/main/scala/io/mediachain/transactor/Dummies.scala:
```diff
@@ -3,7 +3,8 @@ package io.mediachain.transactor
 
 object Dummies {
-  import io.mediachain.transactor.Types._
+  import io.mediachain.types.Datastore
```
this is redundant

yusefnapora replied:
ah, good catch, thanks

yusefnapora committed:
rm redundant import

vyzo commented:
:+1:

parkan commented on transactor/src/main/scala/io/mediachain/transactor/StateMachine.scala:
```diff
@@ -97,15 +98,15 @@ object StateMachine {
             case (EntityChainReference(chain), EntityChainCell(entity, xchain, meta)) => {
               if (checkUpdate(ref, chain, entity, xchain)) {
-                val newcell = EntityChainCell(ref, chain, meta)
+                val newcell = cell.cons(chain)
```
in scala `cons` is called `::`

parkan commented on transactor/src/test/scala/io/mediachain/transactor/CborSerializationSpec.scala:
```diff
@@ -43,12 +55,59 @@ object CborSerializationSpec extends BaseSpec {
+    val entityUpdateCell = EntityUpdateCell(
+      entity = multihashRef("foo"),
+      chain = Some(multihashRef("bar")),
+      meta = Map("favoriteColor" -> CString("Blue! No, Yellow!!!"))
```
it's really fun to come up with these but I would stick them into a list and pull one out at random for each test, similar to how we do it in the L-SPACE fixtures https://github.com/mediachain/L-SPACE/blob/master/l_space/src/test/scala/io/mediachain/GraphFixture.scala#L47

I would also inject some unicode in here

yusefnapora replied:
that's a good call.  Do you think it's worth using scalacheck to generate these?

parkan replied:
I looked into scalacheck generators but they don't seem to be well suited for this task. I think the ad-hoc bags of words are OK for now; we can swap them out for being backed by more sophisticated generators later on.
